main.ml

```diff
@@ -550,50 +561,74 @@ let process root_dir name =
     let args = Args.SR.Probe.request_of_rpc args in
     let name = args.Args.SR.Probe.queue in
     let device_config = args.Args.SR.Probe.device_config in
-    begin match List.find device_config ~f:(fun (k, _) -> k = "uri") with
-    | None ->
-      Deferred.Result.return (R.failure (missing_uri ()))
-    | Some (_, uri) ->
+    compat_uri device_config ~f:(fun device_config ->
       let args = Xapi_storage.Volume.Types.SR.Probe.In.make
         args.Args.SR.Probe.dbg
-        uri in
-      let args = Xapi_storage.Volume.Types.SR.Probe.In.rpc_of_t args in
+        device_config in
+      Xapi_storage.Volume.Types.SR.Probe.In.rpc_of_t args)
+    >>>= fun args ->
+    begin
       let open Deferred.Result.Monad_infix in
-      fork_exec_rpc root_dir (script root_dir name `Volume "SR.probe") args Xapi_storage.Volume.Types.SR.Probe.Out.t_of_rpc
+      let scriptname = script root_dir name `Volume "SR.probe" in
+      fork_exec_rpc root_dir scriptname args Xapi_storage.Volume.Types.SR.Probe.Out.t_of_rpc
       >>= fun response ->
-      let srs = List.map ~f:(fun sr_stat -> sr_stat.Xapi_storage.Volume.Types.sr, {
-        Storage_interface.name_label = sr_stat.Xapi_storage.Volume.Types.name;
-        name_description = sr_stat.Xapi_storage.Volume.Types.description;
-        total_space = sr_stat.Xapi_storage.Volume.Types.total_space;
-        free_space = sr_stat.Xapi_storage.Volume.Types.free_space;
-        clustered = sr_stat.Xapi_storage.Volume.Types.clustered;
-        health = match sr_stat.Xapi_storage.Volume.Types.health with
-          | Xapi_storage.Volume.Types.Healthy _ -> Healthy
-          | Xapi_storage.Volume.Types.Recovering _ -> Recovering
-          ;
-      }) response.Xapi_storage.Volume.Types.SR.Probe.Out.srs in
-      let uris = response.Xapi_storage.Volume.Types.SR.Probe.Out.uris in
-      let result = Storage_interface.(Probe { srs; uris }) in
-      Deferred.Result.return (R.success (Args.SR.Probe.rpc_of_response result))
+        let pp_probe_result () probe_result =
+          Xapi_storage.Volume.Types.SR.Probe.Out.rpc_of_t [probe_result] |>
+          Jsonrpc.to_string
+        in
+        response
+        |> List.map ~f:(fun probe_result ->
+          let uuid = List.Assoc.find probe_result.Xapi_storage.Volume.Types.configuration ~equal:String.equal "sr_uuid" in
+          let open Deferred.Or_error in
+          let smapiv2_probe ?sr_info () =
+            { configuration=probe_result.configuration; complete=probe_result.complete; sr=sr_info; extra_info=probe_result.extra_info }
+          in
+          match probe_result.Xapi_storage.Volume.Types.sr, probe_result.Xapi_storage.Volume.Types.complete, uuid with
+          | _, false, Some _uuid ->
+              errorf "A configuration with a uuid cannot be incomplete: %a" pp_probe_result probe_result
+          | Some sr_stat, true, Some _uuid ->
+              let sr_info = {
```

Comment:
This could be made a little more compact using `Module.{ x = y; }` but I'm not requesting this change: 
```
let sr_info = Xapi_storage.Volume.Types.{
  name_label = sr_stat.name;
  sr_uuid = sr_tat.uuid;
  ...
}
```